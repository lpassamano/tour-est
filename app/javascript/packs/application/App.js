import React, { Component } from "react";
import { Router, Link, navigate } from "@reach/router";

import LoginForm from "./components/LoginForm";
import CreateAccountForm from "./components/CreateAccountForm";
import StaffUserDashboard from "./components/StaffUserDashboard";
import CreateTourForm from "./components/CreateTourForm";
import TourContainer from "./components/TourContainer";
import api from "./api";

const INITIAL_STATE = {
  currentStaffUser: null,
  tours: {
    data: null,
    isFetching: false
  },
  tourView: {
    data: null
  }
};

class App extends Component {
  state = INITIAL_STATE;

  componentDidMount() {
    this.authenticateStaffUser();
  }

  authenticateStaffUser = async () => {
    const result = await api.authenticateStaffUser();

    if (result.ok) {
      this.setState({ currentStaffUser: result.data });
    }

    return result;
  };

  loginStaffUser = async (username, password) => {
    const result = await api.login(username, password);

    if (result.ok) {
      return this.authenticateStaffUser();
    }

    return result;
  };

  logoutStaffUser = event => {
    event.preventDefault();
    api.removeAuthToken();

    this.setState(INITIAL_STATE, () => navigate("/login"));
  };

  registerStaffUser = async attributes => {
    const userResult = await api.createStaffUser(attributes);
    const { username, password } = attributes.user;

    if (userResult.ok) {
      return this.loginStaffUser(username, password);
    }

    return userResult;
  };

  registerTour = async attributes => {
    const tourResult = await api.createTour(attributes);
    return tourResult;
  };

  listTours = async () => {
    this.setState({ tours: { isFetching: true, data: null } });
    const tourList = await api.listTours();
    this.setState({ tours: { isFetching: false, data: tourList.data } });
  };

  showTour = async tourId => {
    // TODO: update getTour so it gets point info too?
    const tour = await api.getTour(tourId);
    this.setState({ tourView: { data: tour.data } });
  };

  render() {
    return (
      <div>
        <h1>Tour-est</h1>
        {this.state.currentStaffUser ? (
          <nav>
            <Link to="/admin">Home</Link>
            <Link to="/tours/new">Create Tour</Link>
            <a href="#" onClick={this.logoutStaffUser}>
              Log Out
            </a>
          </nav>
        ) : (
          <nav>
            <Link to="/login">Sign In</Link>
            <Link to="/sign-up">Create Account</Link>
          </nav>
        )}

        {this.state.currentStaffUser ? (
          <Router>
            <StaffUserDashboard
              path="/admin"
              currentStaffUser={this.state.currentStaffUser}
              tours={this.state.tours}
              listTours={this.listTours}
            />
            <CreateTourForm
              path="/tours/new"
              onCreateTour={this.registerTour}
              currentStaffUser={this.state.currentStaffUser}
            />
            <TourContainer
              path="/tours/:tourId"
              tour={this.state.tourView.data}
              getTour={this.getTour}
            />
          </Router>
        ) : (
          <Router>
            <LoginForm path="/login" onLogin={this.loginStaffUser} />
            <CreateAccountForm
              path="/sign-up"
              onCreateUser={this.registerStaffUser}
            />
          </Router>
        )}
      </div>
    );
  }
}

export default App;
