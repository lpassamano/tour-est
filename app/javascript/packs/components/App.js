import React, { Component } from 'react'
import { Router, Link } from '@reach/router'

import LoginForm from './LoginForm'
import StaffUserDashboard from './StaffUserDashboard'

const App = () => (
  <div>
    <h1>Tour-est</h1>
    <LoginForm />
    <Router>
      <LoginForm path="/sessions" />
      <StaffUserDashboard path="/staff_user" />
    </Router>
  </div>
);

export default App;
