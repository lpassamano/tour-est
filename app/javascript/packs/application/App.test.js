import { App } from "./App";
import React from "react";
import { shallow } from "enzyme";

describe("<App />", () => {
  test("render", () => {
    const authenticateStaffUser = jest.fn();
    const logoutStaffUser = jest.fn();
    const currentStaffUser = {
      id: 1,
      username: "username123",
      cultural_center: { id: 1, name: "MoMA" }
    };

    const component = shallow(
      <App
        authenticateStaffUser={authenticateStaffUser}
        logoutStaffUser={logoutStaffUser}
        currentStaffUser={currentStaffUser}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
