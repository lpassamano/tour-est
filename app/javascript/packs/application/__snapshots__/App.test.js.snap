// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> render 1`] = `
<div>
  <Navigation
    currentStaffUser={
      Object {
        "cultural_center": Object {
          "id": 1,
          "name": "MoMA",
        },
        "id": 1,
        "username": "username123",
      }
    }
    onLogout={[Function]}
  />
  <Router
    className="container"
  >
    <Connect(StaffUserDashboard)
      default={true}
      path="/admin"
    />
    <Connect(ToursList)
      path="/admin/tours"
    />
    <Connect(CreateTourForm)
      path="/admin/tours/new"
    />
    <Connect(UpdateTourForm)
      path="/admin/tours/:tourId/edit"
    />
    <Connect(TourContainer)
      path="/admin/tours/:tourId"
    />
    <Connect(UpdatePointForm)
      path="/admin/tours/:tourId/points/:pointId/edit"
    />
  </Router>
</div>
`;
